### ISSUE TITLE
Evaluating Models on different GPUs

### ISSUE THREAD

**Issue by mrksr** [GPflow/GPflow#294]
I think what I want to do is related to #225 and #162. I can force Tensorflow and therefore GPflow to use a specific GPU by masking the others using `CUDA_VISIBLE_DEVICES`. There are situations however, where masking the GPUs is cumbersome or not possible like the evaluation of multiple models in parallel (I have not tried this) or simply the simultaneous training of models in multiple Jupyter notebooks.

Given the situation of two python kernels for two notebooks and two GPUs, both GPUs are visible to tensorflow in both kernels, which can for example be verified by taking a look at
```python
from tensorflow.python.client import device_lib
device_lib.list_local_devices()
```
How can I manually tell my models which GPU to use? I have tried using `tf.device` like so:
```python
with tf.device('/gpu:1'):
    model = GPflow.sgpr.SGPR(...)
    model.optimize()
```
This does not seem to do the job however, since all calculations still happen on `/gpu:0`. How do I specify the GPU correctly?

**Comment by alexggmatthews** [comment]
@mrksr sorry for the delay getting on to this a lot of us are currently at the NIPS conference #290 . Hopefully there will be more people on call next week.

**Comment by icouckuy** [comment]
GPflow creates graphs in optimize and the autoflow methods. I'm fairly sure the device context is bound to the graph hence it doesn't work. You could try a previous version of GPflow, before the changes of Hensman.

I would say:
model = GPflow.sgpr.SGPR(...)
    with model.graph(): 
      with tf.device('/gpu:1'):
          model.optimize()
could work, but the graph is only created in optimize.

I'm an advocate of having one graph per model, using tf.Variables really as a persistent parameter as it is meant to. This would simplify a lot of things for us too (for marginalizing out mu, tensorboard, etc.), but I acknowledge this is a rather large change (e.g., the whole Param/paramz paradigm would be obsolute and the scikit optimizers will be [a lot?] slower).

**Comment by jameshensman** [comment]
@mrksr yeah, GPflow is not really 'device aware' yet, and I'm not sure how it should be done.

A hack for the moment would be to modify the model class so that the build_likelihood function is called within the `with` statement you specified.  Perhaps try that and let us know if it works?

**Comment by jameshensman** [comment]
@icouckuy 

>I'm an advocate of having one graph per model

If you have such an implementation on a fork somewhere, I'd be glad to take a look. 

You _really_ don't want to concatenate parameters python side, which is what your suggestion implies.  My guess is that >90% of GPflow users just want to use the scipy optimizers, at least to start, so python-side parameter handling is a no-no.

**Comment by hughsalimbeni** [comment]
To add to this: I'd be keen to have a multi-gpu version. If all that's needed is to run the whole of a model on a particular GPU it's quite easy. Replacing L122 of model with 

`self._session = tf.Session(graph=self._graph, config=tf.ConfigProto(allow_soft_placement=True))`

and then adding 

`with tf.device('/gpu:1'):`

at L126 does the trick, and this could easily passed in through the call to optimize. Something similar would work with AutoFlow. I'd be happy to implement this if you think it's worth it, but as this is only useful when using a specific GPU on a multiple GPU machine, it might be a rather specialised use case (my own use is simply to avoid clashing with other users on a shared machine)

 Running different parts of the model on different GPUs is much harder, I think. For one thing, calling tf.device after tf_mode doesn't work so all the param code would need to be re-written. c.f. https://github.com/ICL-SML/gptf
